testutils/src/main/java/com/microsoft/identity/internal/testutils/nativeauth/api/TemporaryEmailService.kt

```diff
@@ -57,32 +67,47 @@ class TemporaryEmailService {
      * To cater for email send and arrival delays, this is retried several times.
      */
     fun retrieveCodeFromInbox(emailAddress: String): String {
-        var retrievedMailbox = api.retrieveMailbox(emailAddress)
-
-        var mailBoxEmpty = retrievedMailbox.isEmpty()
+        var validCodeRetrieved = false
         var count = 0
-        while (mailBoxEmpty && count < 3) {
-            // Wait 3 seconds
-            Thread.sleep(3000)
-
-            // Attempt to retrieve mailbox again.
-            retrievedMailbox = api.retrieveMailbox(emailAddress)
-            mailBoxEmpty = retrievedMailbox.isEmpty()
-
-            // Max 3 retries
-            count++
+        var latestEmailId: String?
+        var otpValue = ""
+        var apiException: Exception? = null
+
+        while (count < numberOfRetries && !validCodeRetrieved) {
+            try {
+                //wait before calling the email endpoint
+                Thread.sleep(retryDelayMs)
+
+                //API returns dates in the UTC timezone, so system time should also be converted
+                val currentTime = ZonedDateTime.ofInstant(Instant.now(), ZoneOffset.UTC)
+
+                val inboxEmails = api.retrieveMailbox(emailAddress)
+                val newEmailId = inboxEmails
```

Comment:
You may want to consider checking if the new email ID is different than the previous one, as there is no purpose in retrying with the same email and thus same OTP.

Reply:
Done.